✨ Load config for top-level module ONLY

ErisDS opened the pull request:
This PR fixes the issues in #3 as best as I can determine.

This PR is a breaking change to the API, meaning it will need to be shipped as ignition 2.0.0!
- By wrapping the config loading into a caching function, we ONLY load nconf when it is needed.
- So far, there is no case where a submodule requires configuration. Therefore, we don't support this usecase
- If we needed to support this usecase, it could be done by passing an argument to the `config()` function, and effectively "naming" the config instance. We could then cache more than one config object. This seems like way too much for a usecase we don't need.
- Because we are no longer EVER loading nconf from within a subdependency, we no longer need to do extra work to handle where the config comes from - it comes from `process.cwd`. 


---

closes #3
- We call config() to initialise the configuration, so that subdependencies don't needlessly call nconf
- The most common first-caller, is `lib/server.js`
- As submodules no longer try to load config, we can use `process.cwd()` again
- This will only work if sub-dependencies DO NOT require config (currently always true)

ErisDS committed:
✨ Load config for top-level module ONLY

closes #3

- We call config() to initialise the configuration, so that subdependencies don't needlessly call nconf
- The most common first-caller, is `lib/server.js`
- As submodules no longer try to load config, we can use `process.cwd()` again
- This will only work if sub-dependencies DO NOT require config (currently always true)

kirrg001 commented:
> So far, there is no case where a submodule requires configuration. Therefore, we don't support this usecase

ok agree for now 👍

kirrg001 commented on lib/config/index.js:
```diff
@@ -1,37 +1,38 @@
+        defaults = require(path.join(parentPath, 'config.example.json'));
+    }
+
+    config.set('NODE_ENV', env);
```
This won't work, because `config.env()` is called **after** this line.
So if `NODE_ENV` is undefined, `nconf` will always override your value with undefined.

Suggestion: move this line beneath `config.env()`:

```
// CASE: env is not set via env variable
if (config.get('NODE_ENV`))
  config.set('NODE_ENV', 'development')
```

Note: Using an env variable name can be confusing. That's why we use a custom variable in Ghost, see https://github.com/TryGhost/Ghost/blob/master/core/server/config/index.js#L41
Imagine you start Ghost without a NODE_ENV parameter, but the config always contains a value for this parameter. But that is not so important, just a tiny detail.

kirrg001 reviewed (changes requested):
Looks all good. I just added one comment.

ErisDS committed:
🎨 Use `env` as name for local env cache

- NODE_ENV looks like an environment variable, and it may be thought that `config.get('NODE_ENV')` will contain the same value as `process.env.NODE_ENV`.
- By using the name `env` it's clear that this is our own env variable, meaning it is also clearer that this might be calculated from other properties or overidden, rather than being === NODE_ENV

ErisDS commented:
I have updated my original commit to move the statement, and added a second commit that renames the config variable.

As these are 2, separate, breaking API changes, I think it might be a good idea to merge this using "rebase and merge" to keep the 2 changes separate, rather than squashing them?
